packages/salesforcedx-vscode-apex/src/views/testOutline.ts

```diff
@@ -0,0 +1,666 @@
+      }
+    }
+  }
+
+  public getTempFolder(): string {
+    if (vscode.workspace.rootPath) {
+      let tempFolder = ospath.resolve(vscode.workspace.rootPath, 'tmp');
+      while (fs.existsSync(tempFolder)) {
+        tempFolder = tempFolder + crypto.randomBytes(16).toString('hex');
+      }
+      return tempFolder;
+    }
+    return '';
+  }
+
+  public async runApexTests(): Promise<void> {
+    await this.refresh();
+    const tmpFolder = this.getTempFolder();
+    const builder = new ReadableApexTestRunCodeActionExecutor(
+      Array.from(ApexTestOutlineProvider.testStrings.values()),
+      false,
+      tmpFolder,
+      this
+    );
+    const commandlet = new SfdxCommandlet(
+      new SfdxWorkspaceChecker(),
+      new EmptyParametersGatherer(),
+      builder
+    );
+    await commandlet.run();
+  }
+
+  public readJSONFile(folderName: string) {
+    if (vscode.workspace.rootPath) {
+      const fullFolderName = ospath.resolve(
+        vscode.workspace.rootPath,
+        folderName
+      );
+      const jsonSummary = this.getJSONFileOutput(folderName);
+      this.UpdateTestsFromJSON(jsonSummary);
+      this._onDidChangeTreeData.fire();
+      this.eventsEmitter.emit('Delete Folder', fullFolderName);
+    }
+  }
+
+  private getJSONFileOutput(fullFolderName: string): FullTestResult {
+    const files = fs.readdirSync(fullFolderName);
+    let fileName = files[0];
+    for (const file of files) {
+      if (
+        file !== 'test-result-codecoverage.json' &&
+        ospath.extname(file) === '.json' &&
+        file.startsWith('test-result')
+      ) {
+        fileName = file;
+      }
+    }
+    fileName = ospath.join(fullFolderName, fileName);
+    const output = fs.readFileSync(fileName).toString();
+    const jsonSummary = JSON.parse(output) as FullTestResult;
+    return jsonSummary;
+  }
+
+  private UpdateTestsFromJSON(jsonSummary: FullTestResult) {
+    const groups = new Set<ApexTestGroup>();
+    for (const testResult of jsonSummary.tests) {
+      const apexGroupName = testResult.FullName.split('.')[0];
+      const apexGroup = this.apexTestMap.get(apexGroupName) as ApexTestGroup;
+      // Check if new group, if so, set to pass
+      if (apexGroup) {
+        groups.add(apexGroup);
+      }
+      const apexTest = this.apexTestMap.get(testResult.FullName) as ApexTest;
+      apexTest.passed = true;
+      if (apexTest) {
+        apexTest.updateIcon(testResult.Outcome);
+        if (testResult.Outcome === 'Fail') {
+          apexTest.passed = false;
+          apexTest.errorMessage = testResult.Message;
+          apexTest.stackTrace = testResult.StackTrace;
+          apexTest.description =
+            apexTest.stackTrace + '\n' + apexTest.errorMessage;
+        }
+      }
+    }
+    groups.forEach(group => {
+      group.updatePassFailLabel();
+      group.description = this.summarize(jsonSummary.summary, group);
+    });
+  }
+
+  private summarize(summary: TestSummary, group: ApexTestGroup): string {
```

Comment:
Move this into its class in a different file to make it more easily testable.
In general I'm not a fan of just using string concatenation.
Is there a different approach that we can use?